src/Tools/dotnet-dsrouter/Program.cs

```diff
@@ -0,0 +1,122 @@
+                description:    "The diagnostics server ipc address to route. Router accepts ipc connections from diagnostic tools " +
+                                "establishing a new route between runtime and diagnostic tool. If not specified " +
+                                "router will use default ipc diagnostics server path.")
+            {
+                Argument = new Argument<string>(name: "ipcServer", getDefaultValue: () => "")
+            };
+
+        private static Option TcpServerAddressOption() =>
+            new Option(
+                aliases: new[] { "--tcp-server", "-tcps" },
+                description:    "The router TCP/IP address using format [host]:[port]. " +
+                                "Router can bind one (127.0.0.1, [::1], 0.0.0.0, [::], ipv4 address, ipv6 address, hostname) " +
+                                "or all (*) interfaces. Launch runtime using DOTNET_DiagnosticPorts environment variable " +
+                                "connecting router TCP server during startup.")
+            {
+                Argument = new Argument<string>(name: "tcpServer", getDefaultValue: () => "")
+            };
+
+        private static Option RuntimeTimeoutOption() =>
+            new Option(
+                aliases: new[] { "--runtime-timeout", "-rt" },
+                description:    "Automatically shutdown router if no runtime connects to it before specified timeout (seconds)." +
+                                "If not specified, router won't trigger an automatic shutdown.")
+            {
+                Argument = new Argument<int>(name: "runtimeTimeout", getDefaultValue: () => Timeout.Infinite)
+            };
+
+        private static Option VerboseOption() =>
+            new Option(
+                aliases: new[] { "--verbose", "-v" },
+                description:    "Enable verbose logging.")
+            {
+                Argument = new Argument<bool>(name: "verbose", getDefaultValue: () => false)
+            };
+
+        private static int Main(string[] args)
+        {
+            StringBuilder message = new StringBuilder();
+
+            var currentColor = Console.ForegroundColor;
+            Console.ForegroundColor = ConsoleColor.Yellow;
+            Console.WriteLine("WARNING: dotnet-dsrouter is an experimental development tool not intended for production environments." + Environment.NewLine);
+            Console.ForegroundColor = currentColor;
+
+            var parser = new CommandLineBuilder()
+                .AddCommand(IpcClientTcpServerRouterCommand())
+                .AddCommand(IpcServerTcpServerRouterCommand())
+                .UseDefaults()
+                .Build();
+
+            ParseResult parseResult = parser.Parse(args);
+
+            if (parseResult.UnparsedTokens.Count > 0)
+            {
+                ProcessLauncher.Launcher.PrepareChildProcess(args);
+            }
+
+            return parser.InvokeAsync(args).Result;
```

Comment:
Is this actually used? This would need to launch a process in a remote location, e.g., iOS/Android emulator, correct?

Reply:
It can be useful since launching the app normally happens through other tooling like dotnet, mlaunch, msbuild or shell scripts if you use CLI. Having the ability to launch that through the router, making sure it automatically pass along the bound listening address so it can be passed over to application using correct env variables etc, could simplify end user experience. Another benefit would be lifetime management of the process, if we launch through dsrouter and we own the process launching the app meaning we can better control the relationship between the router process and the app process life times. So with this you could do like this (just an example):

dotnet-dsrouter server-server -- xharness apple test \ --app=/path/to/an.app --output-directory=out --targets=ios-simulator-64 --set-env=${DOTNET_DiagnosticPorts}

using above dsrouter will only bind to 12.0.0.1 and a dynamic port, all passed over to launch of xharness --set-env.

Reply:
This also mimics functionality (launching a process through the tool) already available in dotnet-counters and dotnet-trace and reuse the same class (ProcessLauncher) to handle this.